analyses/cell-type-ewings/exploratory_analysis/05-cluster-exploration.Rmd

```diff
@@ -0,0 +1,672 @@
+
+```{r, warning=FALSE, message=FALSE}
+# calculate cluster stability
+stability_df <- sce_list |>
+  purrr::imap(\(sce, id){
+    get_cluster_stability(sce, id, all_cluster_results)
+  }) |>
+  dplyr::bind_rows(.id = "library_id") |>
+  dplyr::mutate(
+    k_value = as.factor(k_value),
+    # make sure that k = 5 comes first
+    k_value = forcats::fct_relevel(k_value, "5", after = 0)
+  )
+
+# plot stability across all values of k
+ggplot(stability_df, aes(x = k_value, y = ari)) +
+  geom_jitter(width = 0.1) +
+  facet_grid(rows = vars(library_id)) +
+  labs(title = "Cluster stability") +
+  stat_summary(
+    aes(group = k_value),
+    color = "red",
+    # median and quartiles for point range
+    fun = "median",
+    fun.min = function(x) {
+      quantile(x, 0.25)
+    },
+    fun.max = function(x) {
+      quantile(x, 0.75)
+    }
+  )
+```
+
+Here we see that cluster stability increases as k increases for both libraries. 
+We see a plateau start to appear around k = 20 and k = 25. 
+
+## Compare clusters to cell types from `SingleR`
+
+Now we will compare the clustering results to the cell type assignments obtained from `SingleR` in the `aucell-singler-annotation.sh` workflow. 
+To compare results we will calculate the Jaccard similarity index between clusters and cell types. 
+We expect that good clustering will line up with the cell type assignments so that there is ~ 1 cell type per cluster. 
+
+For the plots, we will only display the top 7 cell types and all other cells will be lumped together into `All remaining cell types`. 
+
+```{r}
+cluster_classification_df <- all_cluster_results |>
+  # add in classifications from singler
+  dplyr::left_join(classification_df, by = c("library_id", "barcodes")) |>
+  dplyr::mutate(
+    # first grab anything that is tumor and label it tumor
+    # NA should be unknown
+    singler_annotation = dplyr::case_when(
+      stringr::str_detect(singler_annotation, "tumor") ~ "tumor",
+      is.na(singler_annotation) ~ "unknown", # make sure to separate out unknown labels
+      .default = singler_annotation
+    ) |>
+      forcats::fct_relevel("tumor", after = 0),
+    # get the top cell types for plotting later
+    singler_lumped = singler_annotation |>
+      forcats::fct_lump_n(7, other_level = "All remaining cell types", ties.method = "first") |>
+      forcats::fct_infreq() |>
+      forcats::fct_relevel("All remaining cell types", after = Inf),
+    nn_param = forcats::fct_relevel(nn_param, "k_5", after = 0)
+  )
+```
+
+
+```{r, fig.height=10, fig.width=7}
+# get heatmap for each library showing cluster assignments vs. singler assignments
+library_ids |>
+  purrr::map(\(id){
+    cluster_celltype_heatmap(
+      cluster_classification_df,
+      id
+    )
+  })
+```
+
+From these plots, it's obvious that K = 5 is over clustering and cell types are spread across a lot of clusters. 
+However, for the other values of K, there don't seem to be any striking differences that are obvious by eye.  
```

Comment:
im intrigued by the macrophage/cluster 9 interaction for `SCPCL000824`, as well as some of what's going on with cluster 10. Looks like things are bouncing back and forth from k=20 -> k=40 for a decent number of the cells?

Reply:
I think the exact numbers of the clusters here can't really be compared between multiple parameters. We don't actually know that the cells that were in cluster 9 when we cluster with one value of k are the same cells that are in cluster 9 or 10 in another value of k. What I see is there is a cluster that has both macrophages and some cells from the "All remaining cell types" class. With k=25, this is cluster 10 and with k=30 this is cluster 9. 

That being said I do agree that maybe k=30 is a little bit better for SCPCL000824. Mostly because of the increase in purity and increase in stability that is shown between K=25 to 30.

Reply:
> I think the exact numbers of the clusters here can't really be compared between multiple parameters. We don't actually know that the cells that were in cluster 9 when we cluster with one value of k are the same cells that are in cluster 9 or 10 in another value of k.

Haha of course.. Friday afternoon brain 😄

Reply:
> That being said I do agree that maybe k=30 is a little bit better for SCPCL000824. Mostly because of the increase in purity and increase in stability that is shown between K=25 to 30.

Do you plan to run these libraries again through the template notebook? If so, I'd leave this one is, and just indicate in the text that 30 could be good too, and save changes for later. Otherwise if this is it for these libraries, I'd just go ahead and use 30.